# Ross Downey - Machine Learning & Statistics

# ATU - Winter 2022

## Fisher's Lady Drinking Tea Problem - Exercises

### Exercise 1.1
***
<br><br>
__The above gives about a 1.5% chance of randomly selecting the correct cups. Calculate the minimum number of cups of tea required to ensure the probability of randomly selecting the correct cups is less than or equal to 1%.__


In [1]:
# Importing math module
import math as math

In [2]:
# The number of permutations of selecting four cups from 8 choices
# Ref https://www.w3schools.com/python/ref_math_comb.asp
math.comb(8,4)

70

In [3]:
# Therefore, the probability of selecting four cups from 8 is:
round(100/(math.comb(8,4)),3)
# using round function for better display of result and multiply by 100 for % value

1.429

If we wanted to increase the number of cups, and consequently the number of cups of tea made with milk first, we could do the following:
***

In [4]:
# Increasing the experiment to 10 cups of tea in total, with 5 of them having been made with milk first
# The number of permutations of selecting five correctly are:
math.comb(10,5)

252

In [5]:
# The probabilityof seelcting 5 cups from 10 is:
round(100/(math.comb(10,5)),3)

0.397

From this it can be stated that the probability of selecting 5 correctly from 10 is much lower than selecting 4 correctly from 8 (0.397% compared to 1.429%). A value somewhere between 4 out of 8, and 5 out of 10 will yield a probability of less than or equal to 1%.
***

In [6]:
# Try 4 cups out of 9
round(100/(math.comb(9,4)),3)

0.794

Adding an extra cup to the test, but maintaining the same number of cups of tea made with milk first yields a probability of approx. 0.8% which is less than  or equal to 1%. This is the correct answer to the initial question.
***

__Bonus: How many would be required if you were to let the taster get one cup wrong while maintaining the 1% threshold?__

To answer this question we need to replace the second digit in the "math.comb" python function with a 3. This indicates that the lady in question needs to only select three of the cups of tea that were made with milk being added first.

In [7]:
# Initially leave the total number of cups at 8 and change the number of correct answers made to 3 
# i.e. "8 choose 3" in mathematical terms

round(100/(math.comb(8,3)),3)

1.786

This implies that the probability of the lady choosing 3 correct cups from eight is approx 1.8%.<br>
In order to answer this question we need a probability in the region of 1%
***

In [8]:
# Try 9 cups

round(100/(math.comb(9,3)),3)


1.19

In [9]:
# And 10 cups

round(100/(math.comb(10,3)),3)

0.833

Although both 9 and 10 cups give an answer in the 1% region the one closest to our original answer of 0.794% is using 10 cups in total.<br>
The number of cups required to yield a probability of 1% with the taster getting one wrong is 10.
***

__Use scipy's version of Fisher's exact test to simulate the Lady Tasting Tea problem.__

In [10]:
# Importing scipy stats

import scipy.stats as ss

Scipy is an open source Python library that is used to perform a wide variety of scientific and technical computing tasks, including statistical operations. <br>
Ref: https://en.wikipedia.org/wiki/SciPy <br>
Ref: https://docs.scipy.org/doc/scipy/reference/stats.html <br>
***
<br>
Fisher's exact test is used to analyse categorical datasets that are classified in two different ways. <br>
In this case our two classifications are either the tea was made with milk first, or the tea was made with milk added second. <br>
The result yielded is called a "P-value" which indicates the difference from a null hypothesis. <br>
The null hypothesis in this case would be that the lady is completely guessing and can't tell the difference between the cups of tea. <br>

***
The fisher exact table below is entered in a 2-Dimensional manner.<br>
The first square brackets [4,0] indicates the likelihood of the lady selecting 4 correct cups of tea which had milk added first.<br>
The second square brackets [0,4] is the other extreme where the lady does not select any correct cup of tea. <br>
The data inbetween is also plotted i.e [1,3], [2,2], [3,1] etc.<br>
This is where the lady selects 1 correct, 2, correct, 3 correct etc. <br>


In [11]:
# using scipy stats
# The P - value below is 0.0285, double the probability calculated previously.
# As this is a 2 sided test this gives the probabaility at both ends of the plot i.e. double.

ss.fisher_exact([[4,0],[0,4]])

(inf, 0.028571428571428536)

In [12]:
# Adapted from official scipy documentation
# Ref: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.fisher_exact.html
# Hypergeometric is a discrete random variable where the data is chosen from a bin
# Ref: https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.hypergeom.html#scipy.stats.hypergeom

from scipy.stats import hypergeom
import numpy as np
table = np.array([[4, 0], [0, 4]])
M = table.sum()
n = table[0].sum()
N = table[:, 0].sum()
start, end = hypergeom.support(M, n, N)
hypergeom.pmf(np.arange(start, end+1), M, n, N)

array([0.01428571, 0.22857143, 0.51428571, 0.22857143, 0.01428571])

The above results are an indication of the probabilities of each outcome.<br>
The probability of selecting all four cups correctly is approx. 1.4% (0.014 * 100)<br>
The probability of selecting 3 out of 4 cups correctly is 22.85%, and so on <br>
This is a more efficient way of calculating the probability of all possible outcomes. <br>
This could be done using the math.comb function but this would be more labour intensive
***

The evaluation of these results is typically done by assessing the p-value and comparing to 0 and 0.05. <br>
The closer the value to zero the more likely that the null hypothesis can be rejected. <br>
In this case the null hypothesis is that the lady can't distinguish between the two types of tea. <br>
The 0.05 value is an arbritrary value (5% probability) that there is some statistical significance in the data. <br>
As previously stated, the lady has a 1.4% probability (1 in 70) of selecting the correct cups at random without any tasting. <br>
Fisher's Lady Tasting Tea Problem has now been simulated using scipy.stats.
***



## Exercise 1.2 - Example code given in scipy stats for independent t-test
Taking the example code from the scipy stats literature, adding comments and optimizing. <br>
[Scipy Stats Documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_ind.html)

In [13]:
from scipy import stats
import numpy as np
rng = np.random.default_rng(123456)
# Importing scipy stats, numpy, and a random number generator to generate large arrays to test

In [14]:
# First example is a t-test on two sets of data with the same mean
# In this case two variables are created using the same conditions of
# mean (loc) = 5, standard seviation(scale) = 10 and equal number of values (size) = 500
# The rng is the same for both  but there will be different individual values
# As the default "equal_var" is used (assuming equal population variances)
rvs1 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)
rvs2 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)
stats.ttest_ind(rvs1, rvs2)


Ttest_indResult(statistic=2.2057925333835944, pvalue=0.027625806725740155)

In [15]:
# In this case the population variances are assumed to be different.
# Welch's t-test is applied here
stats.ttest_ind(rvs1, rvs2, equal_var=False)

Ttest_indResult(statistic=2.205792533383595, pvalue=0.027625815018997385)

In [16]:
# In this example the variance is changed by adjusting the scale (std. dev.)
rvs3 = stats.norm.rvs(loc=5, scale=20, size=500, random_state=rng)
stats.ttest_ind(rvs1, rvs3)

Ttest_indResult(statistic=3.840207944553338, pvalue=0.00013067828674517335)

In [17]:
# Different standard deviations  but population variance also assumed to be different, Welch's t-test
stats.ttest_ind(rvs1, rvs3, equal_var=False)

Ttest_indResult(statistic=3.840207944553338, pvalue=0.00013359292613836685)

In [18]:
# Size of second sample reduced from 500 to 100, std. dev. maintained at 20
# Different sample sizes yield different results for the two t-tests (equal vs. unequal variance t-tests)
rvs4 = stats.norm.rvs(loc=5, scale=20, size=100, random_state=rng)
stats.ttest_ind(rvs1, rvs4)

Ttest_indResult(statistic=0.5880566529111328, pvalue=0.5567162745903518)

In [19]:
# Assumed different variances (Welch's t-test)
stats.ttest_ind(rvs1, rvs4, equal_var=False)

Ttest_indResult(statistic=0.41003925126190405, pvalue=0.682561083538654)

In [20]:
# Different means, variance and size
rvs5 = stats.norm.rvs(loc=8, scale=20, size=100, random_state=rng)
stats.ttest_ind(rvs1, rvs5)

Ttest_indResult(statistic=-2.2593701894479685, pvalue=0.024219519436319857)

In [21]:
# Assumed different variances (Welch's t-test)
stats.ttest_ind(rvs1, rvs5, equal_var=False)

Ttest_indResult(statistic=-1.421609763450849, pvalue=0.15803124827053963)

In [22]:
# Additional permutations of the same arrays leads to a more accurate t-test
stats.ttest_ind(rvs1, rvs5, permutations=10000,random_state=rng)

Ttest_indResult(statistic=-2.2593701894479685, pvalue=0.023)

In [23]:
a = (56, 128.6, 12, 123.8, 64.34, 78, 763.3)
b = (1.1, 2.9, 4.2)

In [24]:
# Trimming, in this case 20%, from the array to remove extreme outliers
# In this case the extreme outlier is 763.3 at the end of array a, array b is not affected
stats.ttest_ind(a, b, trim=.2)

Ttest_indResult(statistic=3.4463884028073513, pvalue=0.01369338726499547)

### Code Improvements
Cleaner code from the scipy stats example above to make it easier to read for the user
***

In [25]:
# Adding all arrays to one cell and adding appropriate variable names to identify easier
arr1 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)
arr2 = stats.norm.rvs(loc=5, scale=10, size=500, random_state=rng)
arr3_sd = stats.norm.rvs(loc=5, scale=20, size=500, random_state=rng)
arr4_size = stats.norm.rvs(loc=5, scale=20, size=100, random_state=rng)
arr5_all = stats.norm.rvs(loc=8, scale=20, size=100, random_state=rng)
to_be_trimmed = (56, 128.6, 12, 123.8, 64.34, 78, 763.3)
not_trimmed = (1.1, 2.9, 4.2)

In [26]:
# Seeting all tests as variables to easily identify what they are
Same_Means = stats.ttest_ind(arr1, arr2)
Same_Means_Welch = stats.ttest_ind(arr1, arr2, equal_var=False)
Diff_Var = stats.ttest_ind(arr1, arr3_sd)
Diff_Var_Welch = stats.ttest_ind(arr1, arr3_sd, equal_var=False)
Diff_Sizes = stats.ttest_ind(arr1, arr4_size)
Diff_Sizes_Welch = stats.ttest_ind(arr1, arr4_size, equal_var=False)
Diff_All = stats.ttest_ind(arr1, arr5_all)
Diff_All_Welch = stats.ttest_ind(arr1, arr5_all, equal_var=False)
Perm =  stats.ttest_ind(arr1, arr5_all, permutations=10000,random_state=rng)
Trim = stats.ttest_ind(to_be_trimmed, not_trimmed, trim=.2)


In [27]:
# Outputting all test results below
# Variable names negates the need to comment each one for the user
print (Same_Means)

Ttest_indResult(statistic=-0.13834707995597603, pvalue=0.8899940276456497)


In [28]:
print(Same_Means_Welch)

Ttest_indResult(statistic=-0.13834707995597603, pvalue=0.8899940613883501)


In [29]:
print(Diff_Var)

Ttest_indResult(statistic=1.8826548992010481, pvalue=0.06003784807179113)


In [30]:
print(Diff_Var_Welch)

Ttest_indResult(statistic=1.8826548992010486, pvalue=0.06014736330234177)


In [31]:
print(Diff_Sizes)

Ttest_indResult(statistic=0.04256861985712979, pvalue=0.9660596197483844)


In [32]:
print(Diff_Sizes_Welch)

Ttest_indResult(statistic=0.03089638281636195, pvalue=0.9754062428844393)


In [33]:
print(Diff_All)

Ttest_indResult(statistic=-2.0474983918820535, pvalue=0.04104554120659462)


In [34]:
print(Diff_All_Welch)

Ttest_indResult(statistic=-1.2534205566732703, pvalue=0.21279380410311588)


In [35]:
print(Perm)

Ttest_indResult(statistic=-2.0474983918820535, pvalue=0.0399)


In [36]:
print(Trim)

Ttest_indResult(statistic=3.4463884028073513, pvalue=0.01369338726499547)


All results are now outputted together using variables that are self-explanatory for easy analysis.
***

# End
***
***